tests/src/whisk/core/database/test/MultipleReadersSingleWriterCacheTests.scala

```diff
@@ -0,0 +1,215 @@
+        // bunch of write inhibits
+        //
+        val inhibits = doCRUD("CONCURRENT CRUD to different keys", { i => "foop_" + i })
+        inhibits.nReadInhibits.get should be(0)
+        inhibits.nWriteInhibits.get should not be (0)
+
+        //
+        // after the first iter, the keys already exist, so the first read
+        // should be cached, resulting in the writes proceeding more
+        // smoothly this time, thus inhibiting some of the second reads
+        //
+        for (i <- 1 to nIters - 1) {
+            doCRUD("CONCURRENT CRUD to different keys", { i => "foop_" + i })
+                .nReadInhibits.get should not be (0)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys" in {
+        for (i <- 1 to nIters) {
+            doCRUD("CONCURRENT CRUD to shared keys", sharedKeys)
+                .nWriteInhibits.get should not be (0)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys (zero latency)" in {
+        var someInhibits = false
+        for (i <- 1 to nIters) {
+            someInhibits = doCRUD("concurrent CRUD to shared keys (zero latency)", sharedKeys, 0)
+                .someInhibits
+        }
+        someInhibits should not be (false)
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys (short latency)" in {
+        for (i <- 1 to nIters) {
+            doCRUD("concurrent CRUD to shared keys (short latency)", sharedKeys, 10)
+                .someInhibits should be(true)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys (medium latency)" in {
+        for (i <- 1 to nIters) {
+            doCRUD("concurrent CRUD to shared keys (medium latency)", sharedKeys, 100)
+                .someInhibits should be(true)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys (long latency)" in {
+        for (i <- 1 to nIters) {
+            doCRUD("CONCURRENT CRUD to shared keys (long latency)", sharedKeys, 5000)
+                .nWriteInhibits.get should not be (0)
+        }
+    }
+
+    "the cache" should "support concurrent CRUD to shared keys, with update first" in {
+        for (i <- 1 to nIters) {
+            doCRUD("CONCURRENT CRUD to shared keys, with update first", sharedKeys, 1000, false)
+                .nWriteInhibits.get should be(0)
+        }
+    }
+
+    def sharedKeys = { i: Int => "foop_" + (i % 2) }
+
+    def doCRUD(
+        testName: String,
+        key: Int => String,
+        delay: Int = 1000,
+        readsFirst: Boolean = true,
+        nThreads: Int = 10): Inhibits = {
+
+        System.out.println();
+        System.out.println(testName);
+
+        val exec = Executors.newFixedThreadPool(nThreads)
+        val inhibits = Inhibits()
+
+        for (i <- 1 to nThreads) {
+            exec.submit(new Runnable { def run() = { doReadWriteRead(key(i), inhibits, readsFirst).go(delay) } })
+        }
```

Comment:
Do we need to manage our own threadpool/concurrency? Can scala handle that for us?